File: solutions/lua/bank-account/1/bank-account.lua
local BankAccount = {}

local closed_account_error  = 'The account is closed.'
local negative_amount_error = 'Amount must be greater than zero.'
local overdraw_error        = 'Amount cannot be greater than the balance.'

function BankAccount:new()
  local bank_account = {}
  setmetatable(bank_account, self)
  self.__index = self

  self.balance_amount = 0
  self.is_open = true

  return bank_account
end

function BankAccount:balance()
  if not self.is_open then error(closed_account_error) end
  return self.balance_amount
end

function BankAccount:deposit(amount)
  if not self.is_open then error(closed_account_error) end
  if amount <= 0 then error(negative_amount_error) end
  self.balance_amount = self.balance_amount + amount
end

function BankAccount:withdraw(amount)
  if not self.is_open then error(closed_account_error) end
  if amount <= 0 then error(negative_amount_error) end
  if amount > self.balance_amount then error(overdraw_error) end

  self.balance_amount = self.balance_amount - amount
end

function BankAccount:close()
  self.is_open = false
end

return BankAccount

File: solutions/javascript/bank-account/1/bank-account.js
export class BankAccount {
  #balance = 0;
  #isOpen  = false;

  open() {
    if (this.#isOpen) {
      throw new ValueError();
    }

    this.#balance = 0;
    this.#isOpen  = true;
  }

  close() {
    if (!this.#isOpen) {
      throw new ValueError();
    }

    this.#isOpen = false;
  }

  deposit(amount) {
    if (!this.#isOpen || amount < 0) {
      throw new ValueError();
    }

    this.#balance += amount
  }

  withdraw(amount) {
    if (!this.#isOpen || amount > this.#balance || amount < 0) {
      throw new ValueError();
    }

    this.#balance -= amount;
  }

  get balance() {
    if (!this.#isOpen) {
      throw new ValueError();
    }

    return this.#balance;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
